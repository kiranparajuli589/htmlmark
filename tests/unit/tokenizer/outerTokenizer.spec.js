const tokenizer = require("../../../lib/tokenizer/outerTokenizer")
const TOKENS = require("../../../lib/tokenizer/tokens")


const commonTokensList = [
  "# Heading One Text",
  "## Heading Two Text",
  "### Heading Three Text",
  "#### Heading Four Text",
  "##### Heading Five Text",
  "###### Heading Six Text",
  "- list item text",
  "1. count item text",
  "<!-- comment item text -->",
  "- [ ] checkbox item text",
  "- [x] checked checkbox item text",
  "> quote item text",
  "> > quote item text",
  "![alt text](image.png)",
]

describe("outer tokenizer", () => {
  describe("newline", () => {
    it.each([
      [""],
      ["\n"],
      ["", "\n"],

    ])("should not include the top empty lines", (lines) => {
      const tokens = tokenizer(lines)
      expect(tokens).toEqual([])
    })
    it.each(["", "\n"])("should include the empty line after some content", (eLine) => {
      const lines = [
        "some plain text",
        eLine,
      ]
      const tokens = tokenizer(lines)
      expect(tokens.length).toBe(2)
      expect(tokens[1].type).toBe(TOKENS.newLine)
    })
    it("should combine multiple consequest new lines to a single one", () => {
      const lines = [
        "some plain text",
        "\n",
        "",
        "some more plain text",
      ]
      const tokens = tokenizer(lines)
      expect(tokens.length).toBe(3)
      expect(tokens[1].type).toBe(TOKENS.newLine)
    })
  })

  describe("codeblock", () => {
    it("should parse the codeblock", () => {
      const lines = [
        "```js",
        "const a = 1",
        "```",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
    it("should parse the codeblock without the language", () => {
      const lines = [
        "```",
        "const a = 1",
        "```",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
    it("should parse the codeblock with multiple lines", () => {
      const lines = [
        "```js",
        "const a = 1",
        "const b = 2",
        "```",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
    it("should parse the codeblock with multiple lines and a newline", () => {
      const lines = [
        "```js",
        "const a = 1",
        "const b = 2",
        "",
        "const c = 3",
        "```",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
    it("should parse multiple consequetive codeblocks", () => {
      const lines = [
        "```js",
        "const a = 1",
        "```",
        "```js",
        "const b = 2",
        "```",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
  })

  describe("common tokens", () => {
    it.each(commonTokensList)("should parse the common tokens", (line) => { 
      const lines = [line]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
    it.each(commonTokensList)("should parse the common tokens with some space before", (line) => {
      const indent = "  "
      const lines = [indent + line]
      const tokens = tokenizer(lines)
      expect(tokens[0].indent).toBe(indent)
    })
  })
  describe("paragraph", () => {
    it("should parse a paragarph as it is", () => {
      const lines = [
        "some plain text",
      ]
      const tokens = tokenizer(lines)
      expect(tokens).toMatchSnapshot()
    })
  })
})