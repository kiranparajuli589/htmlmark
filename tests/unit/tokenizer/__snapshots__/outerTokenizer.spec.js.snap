// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outer tokenizer codeblock should parse multiple consequetive codeblocks 1`] = `
Array [
  Object {
    "language": "js",
    "type": "codeblock",
    "value": "const a = 1",
  },
  Object {
    "language": "js",
    "type": "codeblock",
    "value": "const b = 2",
  },
]
`;

exports[`outer tokenizer codeblock should parse the codeblock 1`] = `
Array [
  Object {
    "language": "js",
    "type": "codeblock",
    "value": "const a = 1",
  },
]
`;

exports[`outer tokenizer codeblock should parse the codeblock with multiple lines 1`] = `
Array [
  Object {
    "language": "js",
    "type": "codeblock",
    "value": "const a = 1
const b = 2",
  },
]
`;

exports[`outer tokenizer codeblock should parse the codeblock with multiple lines and a newline 1`] = `
Array [
  Object {
    "language": "js",
    "type": "codeblock",
    "value": "const a = 1
const b = 2

const c = 3",
  },
]
`;

exports[`outer tokenizer codeblock should parse the codeblock without the language 1`] = `
Array [
  Object {
    "language": null,
    "type": "codeblock",
    "value": "const a = 1",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 1`] = `
Array [
  Object {
    "indent": "",
    "raw": "# Heading One Text",
    "tokens": Object {
      "value": "Heading One Text",
    },
    "type": "heading1",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 2`] = `
Array [
  Object {
    "indent": "",
    "raw": "## Heading Two Text",
    "tokens": Object {
      "value": "Heading Two Text",
    },
    "type": "heading2",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 3`] = `
Array [
  Object {
    "indent": "",
    "raw": "### Heading Three Text",
    "tokens": Object {
      "value": "Heading Three Text",
    },
    "type": "heading3",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 4`] = `
Array [
  Object {
    "indent": "",
    "raw": "#### Heading Four Text",
    "tokens": Object {
      "value": "Heading Four Text",
    },
    "type": "heading4",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 5`] = `
Array [
  Object {
    "indent": "",
    "raw": "##### Heading Five Text",
    "tokens": Object {
      "value": "Heading Five Text",
    },
    "type": "heading5",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 6`] = `
Array [
  Object {
    "indent": "",
    "raw": "###### Heading Six Text",
    "tokens": Object {
      "value": "Heading Six Text",
    },
    "type": "heading6",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 7`] = `
Array [
  Object {
    "indent": "",
    "raw": "- list item text",
    "tokens": Object {
      "value": "list item text",
    },
    "type": "listItem",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 8`] = `
Array [
  Object {
    "indent": "",
    "raw": "1. count item text",
    "tokens": Object {
      "value": "count item text",
    },
    "type": "countItem",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 9`] = `
Array [
  Object {
    "indent": "",
    "raw": "<!-- comment item text -->",
    "tokens": Object {
      "value": "comment item text",
    },
    "type": "comment",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 10`] = `
Array [
  Object {
    "indent": "",
    "raw": "- [ ] checkbox item text",
    "tokens": Object {
      "value": "checkbox item text",
    },
    "type": "checkbox",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 11`] = `
Array [
  Object {
    "indent": "",
    "raw": "- [x] checked checkbox item text",
    "tokens": Object {
      "value": "checked checkbox item text",
    },
    "type": "checkboxChecked",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 12`] = `
Array [
  Object {
    "indent": "",
    "raw": "> quote item text",
    "tokens": Object {
      "value": "quote item text",
    },
    "type": "quote",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 13`] = `
Array [
  Object {
    "indent": "",
    "raw": "> > quote item text",
    "tokens": Object {
      "value": "> quote item text",
    },
    "type": "quote",
  },
]
`;

exports[`outer tokenizer common tokens should parse the common tokens 14`] = `
Array [
  Object {
    "indent": "",
    "raw": "![alt text](image.png)",
    "tokens": Object {
      "alt": "alt text",
      "url": "image.png",
    },
    "type": "image",
  },
]
`;

exports[`outer tokenizer paragraph should parse a paragarph as it is 1`] = `
Array [
  Object {
    "raw": "some plain text",
    "tokens": Object {
      "value": "some plain text",
    },
    "type": "paragraph",
  },
]
`;
